### PR TITLE
DVE-2017-0011: GitBook CDN nameservers return A records on any QTYPE

### PR DESCRIPTION
**Pull request by oerdnj**

### CODE REVIEW

File: 2017/DVE-2017-0011.md
Comment (jelu) [inline 2017/DVE-2017-0011.md:30]: ?

**Review by jelu — CHANGES_REQUESTED** [review]

File: 2017/DVE-2017-0011.md
Comment (oerdnj) [reply 2017/DVE-2017-0011.md:30]: Fixed. And I have also noticed they don't set the 'AA' bit. Anyway I got reponse from their side pretty quick, so this will hopefully get fixed soon.

Comment (jelu) [reply 2017/DVE-2017-0011.md:30]: OK, do you want to wait on that before merge?

**Review by jelu — APPROVED** [review]

**Comment by AaronO** [comment]
@oerdnj I updated your Zendesk Ticket following a deploy last night.

@jelu @oerdnj Do you think you guys could rerun your tests ? With the new changes our DNS now only responds to DNS queries with the following:
* `QClass` is `ANY` or `INET`
* `QType` is `ANY`, `A` or `AAAA`

All other queries get `rcode=NOERROR` with no answer records. Any unexpected errors that happen on our end whilst processing the query now currently return `rcode=SERVFAIL`

**Comment by oerdnj** [comment]
@AaronO There's a great tool called ZoneMaster that can help you with the basics of DNS protocol: https://www.zonemaster.net/test/eb9ee1ee90c5b526

But right now it fails, because you don't handle `NS` queries.

BTW, since you write your DNS in Go, are you aware of CoreDNS (https://coredns.io/)?

**Comment by AaronO** [comment]
@oerdnj Do you mean that if fails because I don't return the SOA on NS queries ? Or that for `dig IN NS cdn.gitbook.com @ns1.gitbook.me`. I should return two `NS` records : `ns1.gitbook.me` and `ns2.gitbook.me` ?

Yep I'm aware of CoreDNS, but the master nodes of our CDN do a few extra things on top of answering DNS queries, such as:
- Tracking healthy edge nodes (edge nodes send heartbeats to the master nodes)
    - So if an edge node is down it can't be returned in DNS replies
- DNS geo-balanced replies (matching the closest healthy edge node to the user)
- Providing an API to control edge nodes (cache clearing, etc ...)
- Provisioning SSL certs
- Auto-clustering via `etcd`

That's all nicely packaged and deployed as a single binary. Maybe we could re-use some code from CoreDNS, but I wouldn't want to use it through it's binary and config files, because that would make things harder to maintain and develop than our current setup.

**Comment by oerdnj** [comment]
> Or that for dig IN NS cdn.gitbook.com @ns1.gitbook.me. I should return two NS records : ns1.gitbook.me and ns2.gitbook.me ?

This ^^^


I am not forcing CoreDNS upon you. It's just ... DNS seems simple at first, but it is hard :), so reusing as much code as possible might prevent some of the mistakes already encountered by others.

**Comment by oerdnj** [comment]
We started and maintain this list exactly for the reason for others to see what tiny little mistakes have other people made (without judging anyone), so the new implementors could avoid making them.

@AaronO Also I would like to express yet again, that I appreciate the fast response and will to fix your DNS implementation!

**Comment by jelu** [comment]
@AaronO If you have special domain names which you would like to control the RRs for there are ways of doing it within the authoritative nameserver software that exists today, Knot-DNS has Lua as do PowerDNS and the PowerDNS team also made dnsdist which can act like a proxy.

**Comment by AaronO** [comment]
Thanks, I'll get that NS fix out sometime later this week.

Sorry if that came across as agressive, CoreDNS is cool, we actually use the developer's underlaying dns library :) So I'll look into using some of CoreDNS' code if possible.

I totally agree, the devil is in the details, I appreciate the community effort to help DNS implementations become standards compliant, hopefully avoiding edge cases for end-users.

**Pull request merged**
